lib/sonar/search.rb

```diff
@@ -2,28 +2,52 @@
 
 module Sonar
   module Search
+
+    # Allow IP queries to be in the form of "1.", "1.2.", "1.2.3.", and "1.2.3.4"
+    IS_IP = /^(\d{1,3}\.|\d{1,3}\.\d{1,3}\.|\d{1,3}\.\d{1,3}\.\d{1,3}\.|\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})$/
+
     # Implemented search query types
-    QUERY_TYPES_MAP = {
-      'certificate' => 'Certificate lookup',
-      'certips'     => 'Certificate to IPs',
-      'rdns'        => 'IP to Reverse DNS Lookup or DNS Lookup to IP',
-      'fdns'        => 'Domains to IP or IPs to Domain',
-      'ipcerts'     => 'IP to Certificates',
-      'namecerts'   => 'Domain to Certificates',
-      'links_to'    => 'HTTP References to Domain',
-      'ports'       => 'Open Ports',
-      'processed'   => 'Open Ports (Processed)',
-      'raw'         => 'Open Ports (Raw)',
-      'sslcert'     => 'Certificate Details',
-      'whois_ip'    => 'Whois (IP)'
-    }
+    QUERY_TYPES = [
+      { name: 'certificate', description: 'Certificate lookup', input: 'sha' },
+      { name: 'certips', description: 'Certificate to IPs', input: 'sha' },
+      { name: 'rdns', description: 'IP to Reverse DNS Lookup or DNS Lookup to IP', input: 'ip' },
+      { name: 'fdns', description: 'Domains to IP or IPs to Domain', input: 'domain' },
+      { name: 'ipcerts', description: 'IP to Certificates', input: 'ip' },
+      { name: 'namecerts', description: 'Domain to Certificates', input: 'domain' },
+      { name: 'links_to', description: 'HTTP References to Domain', input: 'domain' },
+      { name: 'ports', description: 'Open Ports', input: 'ip' },
+      { name: 'processed', description: 'Open Ports (Processed)', input: 'ip' },
+      { name: 'raw', description: 'Open Ports (Raw)', input: 'ip' },
+      { name: 'sslcert', description: 'Certificate Details', input: 'sha' },
+      { name: 'whois_ip', description: 'Whois (IP)', input: 'ip' }
```

Comment:
FWIW, this impacts the output of the `types` subcommand.

Before:


```
$  ./bin/sonar types            
{
    "certificate" => "Certificate lookup",
        "certips" => "Certificate to IPs",
           "rdns" => "IP to Reverse DNS Lookup or DNS Lookup to IP",
           "fdns" => "Domains to IP or IPs to Domain",
        "ipcerts" => "IP to Certificates",
      "namecerts" => "Domain to Certificates",
       "links_to" => "HTTP References to Domain",
          "ports" => "Open Ports",
      "processed" => "Open Ports (Processed)",
            "raw" => "Open Ports (Raw)",
        "sslcert" => "Certificate Details",
       "whois_ip" => "Whois (IP)"
}
```

After:

```
$  ./bin/sonar types
[
    [ 0] {
               :name => "certificate",
        :description => "Certificate lookup",
              :input => "sha"
    },
    [ 1] {
               :name => "certips",
        :description => "Certificate to IPs",
              :input => "sha"
    },
    [ 2] {
               :name => "rdns",
        :description => "IP to Reverse DNS Lookup or DNS Lookup to IP",
              :input => "ip"
    },
    [ 3] {
               :name => "fdns",
        :description => "Domains to IP or IPs to Domain",
              :input => "domain"
    },
    [ 4] {
               :name => "ipcerts",
        :description => "IP to Certificates",
              :input => "ip"
    },
    [ 5] {
               :name => "namecerts",
        :description => "Domain to Certificates",
              :input => "domain"
    },
    [ 6] {
               :name => "links_to",
        :description => "HTTP References to Domain",
              :input => "domain"
    },
    [ 7] {
               :name => "ports",
        :description => "Open Ports",
              :input => "ip"
    },
    [ 8] {
               :name => "processed",
        :description => "Open Ports (Processed)",
              :input => "ip"
    },
    [ 9] {
               :name => "raw",
        :description => "Open Ports (Raw)",
              :input => "ip"
    },
    [10] {
               :name => "sslcert",
        :description => "Certificate Details",
              :input => "sha"
    },
    [11] {
               :name => "whois_ip",
        :description => "Whois (IP)",
              :input => "ip"
    }
]
```